File: 0049-group-anagrams/0049-group-anagrams.py
from collections import defaultdict


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagrams = defaultdict(list)
        
        for s in strs:
            anagrams[count_sort(s)].append(s)
            
        return [*anagrams.values()]

    
def count_sort(s: str) -> Tuple[int]:
    ord_a = ord("a")
    char_counts = [0] * 26
    
    for c in s:
        char_counts[ord(c) - ord_a] += 1
        
    return *char_counts,

File: 0049-group-anagrams/0049-group-anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
const groupAnagrams = function(strs) {
    const anagrams = {/* string -> string[] */};
    
    strs.forEach((s) => {
        const counts = countSort(s);
        
        if (counts in anagrams) {
            anagrams[counts].push(s);
        } else {
            anagrams[counts] = [s];
        }
    });
    
    return Object.values(anagrams)
};

const countSort = function (s) {
    const ordA = 'a'.charCodeAt();
    
    const charCounts = Array(26).fill(0);
    
    for (let i = 0; i < s.length; i++) {
        charCounts[s.charCodeAt(i) - ordA] += 1;
    }
    
    return charCounts.toString()
}